File api

marwing opened the pull request:
fixes #62

cdelledonne commented:
Apologies for the long radio silence, I was on holiday and then again very busy with work. I'll review your PR in the next few days. A huge thanks in advance.

marwing commented:
Great. This probably still has some rough edges and is missing docs. Just let me know what you find. I'm quite busy currently so it may take a while until I can address any issues.

cdelledonne reviewed (changes requested):
Looking great! I left some mostly minor comments and questions, let me know if anything is unclear.

> This probably still has some rough edges and is missing docs.

What do you think we need to document? One thing I would mention in README and docs is that CMake 3.14 is required for build target autocompletion.

One more thing: please rebase onto the latest `master`, and see if we need to add anything to `s:buildsys.Init()`, which is called also when the user jumps to another project (perhaps we need to refresh targets there?)

marwing commented on autoload/cmake/buildsys.vim:
```diff
@@ -375,6 +358,7 @@ function! s:SetCurrentConfig(config) abort
     " Save project's current config and build dir.
     let l:state = {'config': a:config, 'build_dir': l:path}
     call s:state.WriteProjectState(s:buildsys.project_root, l:state)
+    call s:RefreshTargets()
```
Because switching config may change available targets depending on CMake configuration (e.g. a test target that only shows up in the Debug config).

cdelledonne replied:
True, but we don't always need to refresh targets in this function. For instance, the first time we call `s:SetCurrentConfig()` is inside `s:ProcessBuildConfig()`, when we're processing the command-line arguments passed by the user with `:CMakeGenerate`. At that point, the build system does not exist yet, so conceptually there's not point in trying to refresh targets.

I would suggest to explicitly call `s:RefreshTargets()` where really needed (`s:buildsys.Init()` and `s:buildsys.Switch()`). This would also make the logic more clear for someone who reads the code in the future, in my opinion.

marwing replied:
Fair enough

marwing commented on autoload/cmake/buildsys.vim:
```diff
@@ -491,10 +477,8 @@ endfunction
 "         list of available build targets
 "
 function! s:buildsys.GetTargets() abort
-    if len(l:self.targets) == 0
-        call s:RefreshTargets()
-    endif
```
Because the available targets may change from outside factors (e.g. user running CMake from the cli). `buildsys` doesn't have a local version of targets anymore and `s:fileapi.Reparse` is already checking if the index changed. I figured it would be easier and more correct to just let the fileapi handle update detection. If this turns out to be too much overhead this can be removed but since most of the time the cached version should be correct this should be OK.

cdelledonne replied:
I think the overhead would indeed be negligible

marwing commented on autoload/cmake/build.vim:
```diff
@@ -70,6 +71,14 @@ function! s:GenerateQuickfix() abort
+"
+function! s:RefreshTargets() abort
+    let l:path_to_current_config = s:buildsys.GetPathToCurrentConfig()
+    let l:build_dir = s:system.Path([l:path_to_current_config], v:true)
```
Whoops. It's not. Missed this.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+" ==============================================================================
+" Location:    autoload/cmake/cmake-api.vim
```
My bad.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+" ==============================================================================
+" Location:    autoload/cmake/cmake-api.vim
+" Description: Functions for interfacing with the cmake-fila-api(7)
```
My bad. Should have enabled spellchecking in comments.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+
+let s:fileapi = {}
+let s:fileapi.cmake_version_supported = v:false
+
```
We can. But this would duplicate the initial state between here and `Reset()`. Since the others do this too, I'll change it anyways.

cdelledonne replied:
We can skip calling `Reset()` at initialization time now, can't we?

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+" Reset all fileapi state
+"
+function! s:fileapi.Reset() abort
+    let s:fileapi.version = 0
```
Leftover from experimenting with query versioning. Will remove.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+function! s:fileapi.Reset() abort
+    let s:fileapi.version = 0
+    let s:fileapi.lastIndexName = 'unset' " won't compare true
+    let s:fileapi.index = {}
+    let s:fileapi.codemodel = {}
+    let s:fileapi.codemodel.targets = []
```
Having `index` here saves me from returning returning it from `ParseIndex` which already has a return value. It could be a script local variable instead of a class variable though. Otherwise you are correct. Having codemodel as a dict is just a thing I did because we might want to add more keys to it, but this would be easy enough to change in the event we did.

marwing replied:
Missed the `codemodel_file` part. This is the only part used from the index outside the `ParseIndex` function. This once again is a "we may want to add more later". Will change. The entire logic to detect the codemodel path seems weird now. I don't even use the results...

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+"
+function! s:fileapi.Reset() abort
+    let s:fileapi.version = 0
+    let s:fileapi.lastIndexName = 'unset' " won't compare true
```
My bad. The others all are.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+"     build_dir : String
+"         path to current build configuration
+"
+function! s:fileapi.Reparse(build_dir) abort
```
I don't know. I had it as `Parse` before but changed it. Both sort of work but are slightly inaccurate. If you think `Parse` is better I will change it.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+" Find the current index file
+"
+function! s:FindIndexFile(build_dir) abort
+    let l:glob = s:system.Path([a:build_dir] + s:reply_path + ['index'], v:true) . '-*.json'
```
I hadn't thought of that. Although it probably doesn't matter too much in this file since this information is either short lived or non crucial. Will change anyway.

cdelledonne replied:
Actually the result of `s:FindIndexFile()` is not short lived, it gets saved into `s:fileapi.last_index_name`. If, after running this once, we `:cd` into a subfolder of the root folder, and obtain the new relative path to the index file, `s:ParseIndex()` will think the new file name differs from the previous one.

marwing replied:
While the result is not short lived it is only used in a caching manner. Changing directory would essentially just invalidate the cache. Nothing would break because of this (most likely, CMake names its index file with a timestamp down to the millisecond). In any case, it is more correct to save the full path.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,188 @@
+" Reset all fileapi state
+"
+function! s:fileapi.Reset() abort
+    let s:fileapi.version = 0
```
Another thing I missed. Let's see if I ever get the hang of vimscript.

cdelledonne replied:
Not the most sensible language out there (:

marwing commented:
I think I addressed all your comments by either a change or an explanation. I also rebased on master. I don't think we have to do anything special in `s:buildsys.Init()`. You already call `s:SetCurrentConfig` which does the necessary refreshing.

For documentation I think we should at least document the new version requirement and the fact that the file api is used at all. The rest is internals I would say. Also the changelog entry is still missing.

cdelledonne commented on autoload/cmake/buildsys.vim:
```diff
@@ -461,6 +444,7 @@ function! s:buildsys.Generate(clean, argstring) abort
     let l:run_options.callbacks_err = [function('s:RefreshConfigs')]
     let l:run_options.autocmds_pre = ['CMakeGeneratePre']
     call s:terminal.Run(l:command, 'GENERATE', l:run_options)
+    call s:fileapi.Parse(l:optdict.build_dir)
```
Is this call to `Parse()` necessary? We will parse anyway in the `s:RefreshTargets()` callback, won't we?

marwing replied:
Indeed

cdelledonne commented:
Thanks a lot for the quick reaction, back to you with a couple more comments!

cdelledonne reviewed (commented):
The changes look good to me! Once you're done with documentation stuff, I think we're ready to merge.

marwing commented:
I think this should be enough for documentation since this is more of an internal change. Let me know if you would reword anything. Otherwise I would say this is ready.

cdelledonne commented on README.md:
```diff
@@ -14,7 +14,8 @@ nice visual feedback.
 * Visual experience, shows CMake output in a console-like window
 * Slick management of build configurations
-* Autocompletion for build targets, build configurations and tests
+* Autocompletion for build targets, build configurations and tests using the
+  `cmake-file-api(7)`
```
Actually, the file API is only used for build targets. Also, I think this is too low-level of a detail to be mentioned in the feature overview. Perhaps we can revert this and mention the file API in the requirements section? (see comment below)

marwing replied:
Sure.

cdelledonne commented on README.md:
```diff
@@ -24,6 +25,7 @@ nice visual feedback.
 
 * Vim with `+terminal`, or Neovim >= 0.5
 * Under Windows, only Neovim is supported at the moment
+* Target completion requires CMake 3.14 or later
```
E.g. "Target autocompletion uses the `cmake-file-api(7)`, which requires CMake 3.14 or later"

marwing replied:
It's not autocompletion, is it?

cdelledonne replied:
I think just "completion" is enough

cdelledonne commented on README.md:
```diff
@@ -24,6 +25,7 @@ nice visual feedback.
 
 * Vim with `+terminal`, or Neovim >= 0.5
 * Under Windows, only Neovim is supported at the moment
+* Target completion requires CMake 3.14 or later
```
By the way, we should add this line to doc/cmake.txt as well

marwing replied:
Oops. I missed this section is in docs as well. Will add.

marwing replied:
Should I leave [this](https://github.com/cdelledonne/vim-cmake/pull/64/files#diff-b73fc732b2dcef182f1fed830367b06d323bb053305436c27b6c6321d82862b3R140-R141) in then?

cdelledonne replied:
I would leave this in, yes (just add a blank line right after these two lines, forgot to point it out earlier)

cdelledonne commented on autoload/cmake/fileapi.vim:
```diff
@@ -54,6 +54,9 @@ endfunction
     let l:index_path = s:FindIndexFile(a:build_dir)
     if l:index_path ==# ''
+        " TODO: only show this with an existing build tree without api files
+        call s:logger.EchoError('Api response missing, run :CMakeGenerate')
+        call s:logger.LogError('Api response missing, run :CMakeGenerate')
```
The error message is in my opinion too abstract for the user, most people who will get this error won't know what "Api response missing" means. Could we make this message a bit more meaningful to the user? Is this too vague? "List of build targets missing, run `:CMakeGenerate`"

marwing replied:
I know this is not a good message. I just noticed this would be a problem and added something to not forget. I think 'List of build targets' is too vague. I, unfortunately, don't have a better idea myself. The only thing I can think of would be to explicitly name the `cmake-file-api(7)` here too. This is only slightly better, but at least you could google (or `:Man`) what this is.

cdelledonne replied:
I would not get too technical, as all the user cares about is in fact the latter part of the sentence. I also don't have any great suggestions now, so I'll leave this up to you, and we can refine in the future if necessary

marwing replied:
I like to think that people using vim to write software in a language that uses CMake for a buildsystem are at least somewhat interested in technicalities. (Could be wrong though)
As I think 'Api response' is in fact not useful at all I will change this to include the reference to the file api. If we ever find some better message we can change it. I will also explicitly mention that target completion will not work because of this.

cdelledonne reviewed (changes requested):
A couple of rewording suggestions, let me know what you think

cdelledonne commented on autoload/cmake/fileapi.vim:
```diff
@@ -0,0 +1,192 @@
+    if l:index_path ==# ''
+        " TODO: only show this with an existing build tree without api files
+        " and when cmake version new enough
+        let l:warning =
+                    \ 'fileapi: Response from cmake-file-api(7) missing.'
+                    \ . ' Target completion will not work.'
+                    \ . ' Run :CMakeGenerate'
```
I realized this warning gets echoed whatever command you run when there isn't a build system yet. Even if you just run `:CMakeGenerate` without a build system being there yet. Which is not what we want I guess.

Ideally, this warning gets echoed only if the user tries to tab-complete targets before generating a build system, i.e. only when this function is triggered as a consequence of invoking `cmake#GetBuildTargets()` (from autoload/cmake.vim). We should not issue this warning if `s:fileapi.Parse()` is invoked from other code paths.

Would you agree? To circumvent this problem, we could only raise an exception from this function, and silently ignore such exception everywhere except for in `cmake#GetBuildTargets()`, where we do echo the warning. Or something else that has the same effect.

Sorry I'm raising this just now, I hadn't tested this properly.

cdelledonne commented:
Also, given that `s:buildsys.GetTargets()` is only called once now (from `s:GetBuildArgs()` in build.vim), we could get rid of that, and always use `s:fileapi` to retrieve targets.

marwing commented on autoload/cmake/fileapi.vim:
```diff
@@ -163,12 +161,12 @@ endf
     endif
 
-    let l:ret = s:ParseIndex(a:build_dir)
+    let l:ret = s:ParseIndex(a:build_dir, a:throw)
```
I mostly agree, I don't know what I was thinking. I would throw directly from `s:ParseIndex()` however, as this allows us to throw different exception for different errors. If we still name them in a consistent way we still can catch `/vim-cmake_fileapi_.*/` in function in which we just don't care. This also allows us to return the response files array locally instead of using a 'global' variable and an error code.

cdelledonne commented:
The latest version look great! Could I just ask you to update the docstrings of these functions in fileapi.vim? Where applicable, they should include `Params`, `Returns`, and perhaps also `Raises` for those that can raise exceptions.
- `s:ParseIndex()`
- `s:ParseCodemodel()`
- `s:fileapi.Parse()`

After that, I'll just merge, if you agree.

cdelledonne commented:
And by the way, thanks A LOT for taking the time to work on this feature and address my review comments, I appreciate that a lot

marwing commented:
> and perhaps also `Raises` for those that can raise exceptions.

Does this mean for all those from which a vim-cmake exception can escape or only those with the actual throw statement (does it include `s:fileapi.Parse()` or only `s:ParseIndex()`).

> After that, I'll just merge, if you agree.

I do.

cdelledonne commented:
I meant to also document `Raises` in `s:fileapi.Parse()`, given that it's a public function, so that if we are to use it in some other places in the future we know how to treat it. It'd be great if you could apply this last one change. Thanks!

cdelledonne commented:
That was fast! I'm merging

marwing commented:
> That was fast! I'm merging

I may have had this prepared in case this is what you wanted... :)

Anyways, thanks for merging.
